lib/srv/heartbeatv2_test.go

```diff
@@ -0,0 +1,325 @@
+		// with a blocking strategy instead here to avoid dealing w/ backoff that could make the
+		// test need to run longer.
+		select {
+		case <-ctx.Done():
+			return nil, trace.Errorf("context canceled while waiting for next control stream")
+		case stream := <-streamC:
+			return stream, nil
+		}
+	}, hello)
+
+	return &fakeHeartbeatDriver{
+		handle:      handle,
+		streamC:     streamC,
+		pollChanged: 1, // first poll is always true
+	}
+}
+
+// TestHeartbeatV2Basics verifies the basic functionality of HeartbeatV2 under various conditions by
+// injecting a fake driver and attempting to force the HeartbeatV2 into all of its happy and unhappy
+// states while monitoring test events to verify expected behaviors.
+func TestHeartbeatV2Basics(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// set up fake hb driver that lets us easily inject failures for
+	// the diff steps and assists w/ faking inventory control handles.
+	driver := newFakeHeartbeatDriver()
+	defer driver.handle.Close() // driver includes fake handle that needs cleanup
+
+	hb := newHeartbeatV2(driver.handle, driver, heartbeatV2Config{
+		announceInterval: time.Millisecond * 200,
+		pollInterval:     time.Millisecond * 50,
+		fallbackBackoff:  time.Millisecond * 400,
+		testEvents:       make(chan hbv2TestEvent, 1028),
+	})
+	go hb.Run()
+	defer hb.Close()
+
+	// initial state: fallback announce and polling occur, but
+	// no control stream is available yet, so we don't ever
+	// use the control-stream announce. First poll always reads
+	// as different, so expect that too.
+	awaitEvents(t, hb.testEvents,
+		expect(hbv2PollDiff, hbv2FallbackOk, hbv2Start),
+		deny(hbv2FallbackErr, hbv2FallbackBackoff, hbv2AnnounceOk, hbv2AnnounceErr),
+	)
+
+	// verify that we're now polling "same" and that time-based announces
+	// are being triggered (we set the announce interval very short, so these
+	// should be going off a lot).
+	awaitEvents(t, hb.testEvents,
+		expect(hbv2PollSame, hbv2AnnounceInterval),
+		deny(hbv2PollDiff),
+	)
+
+	// set up some fallback errs
+	driver.mu.Lock()
+	driver.fallbackErr = 2
+	driver.mu.Unlock()
+
+	// wait for fallback errors to happen, and confirm that we see fallback backoff
+	// come into effect. we still expect no proper announce events.
+	awaitEvents(t, hb.testEvents,
+		expect(hbv2FallbackErr, hbv2FallbackErr, hbv2FallbackBackoff, hbv2FallbackOk),
+		deny(hbv2AnnounceOk, hbv2AnnounceErr),
+	)
+
+	// confirm we resume non-err fallback calls (gotta check this separately
+	// to establish ordering).
+	awaitEvents(t, hb.testEvents,
+		expect(hbv2FallbackOk, hbv2FallbackOk),
+		deny(hbv2FallbackErr, hbv2AnnounceOk, hbv2AnnounceErr),
+	)
+
+	// make a stream available to the heartbeat instance
+	// (note: we don't need to pull from our half of the stream since
+	// fakeHeartbeatDriverInner doesn't actually send any messages across it).
+	stream := driver.newStream(ctx, t)
+	defer stream.Close()
```

Comment:
Same thing here, can `newStream` use `t.Cleanup` so we don't need to remember to close it?